spring/src/main/kotlin/com/katanox/tabour/spring/config/TabourConfigurer.kt

```diff
@@ -6,66 +6,64 @@ import com.katanox.tabour.configuration.core.tabour
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
-import org.springframework.boot.autoconfigure.AutoConfiguration
-import org.springframework.context.ConfigurableApplicationContext
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.ApplicationContext
+import org.springframework.context.ApplicationListener
 import org.springframework.context.annotation.Bean
 import org.springframework.context.annotation.Lazy
+import org.springframework.context.event.ContextRefreshedEvent
+import org.springframework.stereotype.Component
 import org.springframework.util.ClassUtils
 
 val scope = CoroutineScope(Dispatchers.IO)
 
-@AutoConfiguration
-open class TabourConfigurer {
+class TabourConfigurer
 
-    @Bean
-    @Lazy(false)
-    open fun setupTabour(context: ConfigurableApplicationContext): Tabour {
+@Component
+class ContextRefreshedEventListener(
+    @Value("\${tabour.config.num-of-threads:2}") val threadsCount: Int
+) : ApplicationListener<ContextRefreshedEvent?> {
+    override fun onApplicationEvent(contextRefreshedEvent: ContextRefreshedEvent?) {
+        if (contextRefreshedEvent?.applicationContext != null) {
+            setupTabour(contextRefreshedEvent.applicationContext)
+        }
+    }
+
+    @Bean @Lazy(false) fun tabourBean(): Tabour = tabour { this.numOfThreads = threadsCount }
+
+    private fun setupTabour(context: ApplicationContext) {
         val annotatedBeans: Map<String, Any> =
             context.getBeansWithAnnotation(AutoconfigureTabour::class.java)
 
         // only one class should be annotated with this annotation
-        return if (annotatedBeans.isNotEmpty() && annotatedBeans.size == 1) {
+        if (annotatedBeans.isNotEmpty() && annotatedBeans.size == 1) {
             val mainClass = ClassUtils.getUserClass(annotatedBeans.values.first())
 
-            val beanFactory = context.beanFactory
-
-            val threads =
-                context.environment.getProperty("tabour.config.num-of-threads", Int::class.java)
-                    ?: 2
+            val tabourContainers = context.getBeansOfType(Tabour::class.java)
 
-            constructTabourContainer(mainClass, threads) {
-                beanFactory.getBeansOfType(Registry::class.java).values.toList()
+            if (tabourContainers.size == 1) {
+                maybeLaunchTabour(mainClass, tabourContainers.values.first()) {
+                    context.getBeansOfType(Registry::class.java).values.toList()
+                }
             }
-        } else {
-            tabourBean(emptyList(), 1)
         }
     }
 }
 
-internal fun <T> constructTabourContainer(
-    c: Class<T>,
-    numOfThreads: Int = 1,
+internal fun maybeLaunchTabour(
```

Comment:
instead of starting the name with maybe, we can name it launchTabour (it's a standard to start the func name with a verb) and as a comment, we can explain this function launches tabour only if there are necessary annotations and registries?

Reply:
well `launchTabour` can be misleading because it is not started always but i will change it